cpp/src/parquet/types.cc

```diff
@@ -1630,6 +1669,217 @@ class LogicalType::Impl::Float16 final : public LogicalType::Impl::Incompatible,
 
 GENERATE_MAKE(Float16)
 
+namespace {
+std::string EscapeControl(char c) {
```

Comment:
You're writing the JSON string into a stream anyway, so why not do that here as well?
```suggestion
void EscapeControl(char c, std::ostream& json) {
```